platform/rpc/nats/bus.go

```diff
@@ -0,0 +1,246 @@
+package nats
+
+import (
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/nats-io/nats"
+
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/platform"
+	"github.com/weaveworks/flux/platform/rpc"
+)
+
+const (
+	timeout      = 5 * time.Second
+	presenceTick = 50 * time.Millisecond
+	encoderType  = nats.JSON_ENCODER
+
+	methodPing         = ".Platform.Ping"
+	methodAllServices  = ".Platform.AllServices"
+	methodSomeServices = ".Platform.SomeServices"
+	methodRegrade      = ".Platform.Regrade"
+)
+
+type NATS struct {
+	url string
+	// It's convenient to send (or request) on an encoding connection,
+	// since that'll do encoding work for us. When receiving though,
+	// we want to decode based on the method as given in the subject,
+	// so we use a regular connection and do the decoding ourselves.
+	snd *nats.EncodedConn
+	rcv *nats.Conn
+}
+
+var _ platform.MessageBus = &NATS{}
+
+func NewMessageBus(url string) (*NATS, error) {
+	conn, err := nats.Connect(url)
+	if err != nil {
+		return nil, err
+	}
+	encConn, err := nats.NewEncodedConn(conn, encoderType)
+	if err != nil {
+		return nil, err
+	}
+	return &NATS{
+		url: url,
+		rcv: conn,
+		snd: encConn,
+	}, nil
+}
+
+// Wait up to `timeout` for a particular instance to connect. Mostly
+// useful for synchronising during testing.
+func (n *NATS) AwaitPresence(instID flux.InstanceID, timeout time.Duration) error {
+	timer := time.After(timeout)
+	attempts := time.NewTicker(presenceTick)
+	defer attempts.Stop()
+
+	for {
+		select {
+		case <-attempts.C:
+			if err := n.Ping(instID); err == nil {
+				return nil
+			}
+		case <-timer:
+			return errors.New("presence timeout")
+		}
+	}
+}
+
+func (n *NATS) Ping(instID flux.InstanceID) error {
+	var response PingResponse
+	err := n.snd.Request(string(instID)+methodPing, ping{}, &response, timeout)
+	if err == nil {
+		err = maybeError(response.Error)
+	}
+	return err
+}
+
+type AllServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type SomeServicesResponse struct {
+	Services []platform.Service
+	Error    string
+}
+
+type RegradeResponse struct {
+	Result rpc.RegradeResult
+	Error  string
+}
+
+type ping struct{}
+
+type PingResponse struct {
+	Error string
+}
+
+func maybeError(msg string) error {
+	if msg != "" {
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func maybeString(err error) string {
+	if err != nil {
+		return err.Error()
+	}
+	return ""
+}
```

Comment:
> Well, if they were conversion functions

`stringAsError` is a function from `string` to `error` (`error` includes `error(nil)`).

But anyway, taking the charge of readability seriously, I rewrote a procedure to do the same conversation explicitly inline:

```
func (r *requester) AllServices(ns string, ig flux.ServiceIDSet) ([]platform.Service, error) {
	var response AllServicesResponse
	var err error
	if err = r.conn.Request(r.instance+methodAllServices, rpc.AllServicesRequest{ns, ig}, &response, timeout); err != nil {
		return nil, err
	}
	if response.Error != "" {
		err = errors.New(response.Error)
	}
	return response.Services, err
}
```

Honestly, I do not see how this makes it more readable -- it just adds some noise, in the form of unimportant detail, which must be analysed each time to make sure it's not doing something important (or subtly different to the last such block). I'm not a zealot who wants to factor everything into noun soup. I really think in this instance that putting the conversion into a function and using it in value position makes things easier to follow.